### ISSUE TITLE
BlenderBIM AttributeError (without any using API related with ifc/bim)

### ISSUE THREAD

**Issue by koji-dan** [IfcOpenShell/IfcOpenShell#1632]
Hi, i always appreciate for you great works !

What i want to try is just to import/export ifc file by python console of blender, and i installed BlenderBIM addon into blender.
After installing BlenderBIM, i always meet AttributeError on python script as the followings.
this message is always showing on starting python script, when not only before import/export call but also not using ifc import/export. this might be based on the version combination issue ??

<error message>
Blender 2.93.0 (hash 84da05a8b806 built 2021-06-02 11:34:30)
Read prefs: C:\Users\r00414643\AppData\Roaming\Blender Foundation\Blender\2.93\config\userpref.blend
Exception in module register(): C:\Users\r00414643\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\blenderbim\__init__.py
Traceback (most recent call last):
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\scripts\modules\addon_utils.py", line 386, in enable
    mod.register()
  File "C:\Users\r00414643\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\blenderbim\bim\__init__.py", line 131, in register
    module.register()
  File "C:\Users\r00414643\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\blenderbim\bim\module\model\__init__.py", line 34, in register
    bpy.utils.register_tool(workspace.BimTool, after={"builtin.scale_cage"}, separator=True, group=True)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\scripts\modules\bpy\utils\__init__.py", line 888, in register_tool
    tool_converted = tool_from_class(tool_cls)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\scripts\modules\bpy\utils\__init__.py", line 885, in tool_from_class
    cls._km_action_simple(kc_default, kc, context_descr, tool_def.label, keymap_data)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\scripts\startup\bl_ui\space_toolsystem_common.py", line 471, in _km_action_simple
    km = kc.keymaps.get(km_idname)
AttributeError: 'NoneType' object has no attribute 'keymaps'
Traceback (most recent call last):
  File "C:\Users\r00414643\AppData\Roaming\Blender Foundation\Blender\2.93\scripts\addons\blenderbim\bim\handler.py", line 223, in setDefaultProperties
    if len(bpy.context.scene.DocProperties.drawing_styles) == 0:
AttributeError: 'Scene' object has no attribute 'DocProperties'

<Blender: blender-2.93.2-windows-x64.msi>
windows X64
ver 2.93.2 and also ver 2.93.0

<BlenderBIM addon: blender28-bim-210731-py39-win>
v0.0.210731

**Comment by Moult** [comment]
Ah you may have an old user preferences from an older version of the BlenderBIM Add-on.

The BlenderBIM Add-on executes a few things on start up to make drawing generating easier. This has changed a few times because the drawing system has changed a few times too. Since we're still in alpha, I haven't bothered to write code to migrate old settings, so as a result people with old user startup files kinda need to refresh them.

In short, try clearing your startup settings and it should be fixed.

**Comment by koji-dan** [comment]
Thank you so much for your response !
i tried to clear starup setting as the below, but result is not good for me.

1. clean-install of blender 2.93.2
2. go to Window10 AppData folder (C:\Users\{user folder}\AppData\Roaming\Blender Foundation\Blender\2.93\config)
3. I did not create my own user setting file, then there is only userpref.blend in this folder as blend file.
4. when removing it, i do not find Attribute Error on starting python script , but also BlenderBIM addon is disappeared
5. therefore, i can not use ifc import/export by BlenderBIM as my purpose.

**Comment by Moult** [comment]
Is the BlenderBIM Add-on enabled in your add-ons settings in Edit->Preferences?

**Comment by koji-dan** [comment]
in Edit->Preference window,
when BlenderBIM is not enable(disappeared or gray-down display), i don't see AttributeError on python script, but i cannot use ifc import/export of BlenderBIM.
when BlenderBIM is enable, i see AttributeError on python script, and i see also similar AttributeError on use ifc import/export...

**Comment by Moult** [comment]
I think the first thing to do is to ensure that you do not have any errors with the pure BlenderBIM Add-on and you can import/export from the graphical interface, i.e. File->Import->IFC. Then, we can focus on the errors with your script `OpeConvBlender.py`.

Can you first confirm that the BlenderBIM Add-on works by itself? If so, then we can debug your script, but you may need to post your script for me to see what is happening.

**Comment by koji-dan** [comment]
yes, i agree.
from blender GUI, i tried to import ifc file (File -> Import -> IFC) and expot it as stl file (File -> export -> STL).
all of them are succeeded. but python script are not succeeded as the above error.

---------------------------------------------------
(how to wake up the script)
%> python StartScript.py

---------------------------------------------------
(python script-1:  StartScript.py)
# -*- coding: utf-8 -*-
import sys
import os
import os.path
import subprocess
Cmd = "blender.exe --background --python "
Cmd = Cmd + " OpeConvBlender.py "
subprocess.run( Cmd, shell=True )

---------------------------------------------------
(python script-2:  OpeConvBlender.py)
# -*- coding: utf-8 -*-
import bpy
import os
import sys
import time
bpy.ops.import_ifc.bim(filepath="sample_bim.ifc")
bpy.ops.export_mesh.stl( filepath="sample_bim.stl", filter_glob="*.stl" )

**Comment by Moult** [comment]
@koji-dan thanks, this is a pretty fascinating error and I'm not quite sure how to fix it.

It seems that when Blender is in background mode, the keymaps cannot be updated. As a short term fix, you can delete the lines of code you see here: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/blenderbim/blenderbim/bim/module/model/workspace.py#L17-L32 that I have highlighted (L17-L32).

It seems as though the existence of a `bl_keymap` will cause the add-on to fail to load in background mode.

Pinging @s-leger and @Gorgious56 in case they know how to fix this.

To replicate do:

```
$ cat foo.py
import bpy
bpy.ops.import_ifc.bim(filepath="/home/dion/wall.ifc")
$ blender --background --python foo.py
```

We'll keep this open until we find a proper fix, but I hope the workaround helps you solve your immediate problem.

**Comment by s-leger** [comment]
Easy fix:
```
if bpy.app.background:
    return

```

**Comment by koji-dan** [comment]
thank you so much !
i could confirm no error without --backgournd in command line according to your comment.
